// wise-api/lib/resources/invoice/actions/InvoiceActions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const AppError_1 = __importDefault(require("../../../common/errors/AppError"));
const resource_1 = __importDefault(require("../../../common/resource"));
class Invoice extends resource_1.default {
    constructor(token, url, domain) {
        super(token, url);
        this.url = `${this.url}/invoices`;
        this.domain = domain;
    }
    create(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    get(invoiceId) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.get(`${this.url}/${invoiceId}`, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    list(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/list`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getOrgUnitView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/orgunit`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getOrgUnitAndServiceView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/orgunit/service`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getDayView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/day`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getServiceView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/service`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getServiceAndDayView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/service/day`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getUserView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/user`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getUserAndServiceView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/user/service`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    getDetailView(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/detail`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    update(invoiceId, data) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.request.patch(`${this.url}/${invoiceId}`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    delete(invoceId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.request.delete(`${this.url}/${invoceId}`, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
}
exports.default = Invoice;
