// wise-api/lib/resources/room/actions/RoomActions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const AppError_1 = __importDefault(require("../../../common/errors/AppError"));
const resource_1 = __importDefault(require("../../../common/resource"));
class Room extends resource_1.default {
    constructor(token, url) {
        super(token, url);
        this.url = `${this.url}/rooms`;
    }
    create(data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    get(room) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.get(`${this.url}/${room}`, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { console.log(err); throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    list(filters) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/list`, filters, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    update(room, data) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.request.patch(`${this.url}/${room}`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    delete(room) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.request.delete(`${this.url}/${room}`, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    open(room, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/${room}/OPEN`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    join(room, data) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request.post(`${this.url}/${room}/JOIN`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .then((res) => res.data)
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
    close(room, data) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.request.post(`${this.url}/${room}/CLOSE`, data, { headers: { Authorization: this.token ? `Bearer ${this.token}` : '' } })
                .catch((err) => { throw new AppError_1.default(err.response.data, err.response.status); });
        });
    }
}
exports.default = Room;
